### PR TITLE
feat(task): task add feature

### PR DESCRIPTION
**Pull request by RukaiYu**

### CODE REVIEW

File: Server/API/Controllers/TaskController.cs [hunk 1]
```diff
@@ -20,14 +24,21 @@ public ActionResult<string> GetTask()
...
         }
 
         // PUT api/values/5
-        [HttpPost("{id}")]
-        public void AddTask(int id, [FromBody] Core.Task task)
+        [HttpPost]
+        public ActionResult AddTask([FromBody][Required] Core.Task task)
         {
-            //do nothing for now
+            if (task.Name == "!@#$%")
```
Comment (LingForCC) [inline Server/API/Controllers/TaskController.cs]: The requirement is to limit all special characters, which means that the name can only be English characters and numbers.  I think we need a regular expression here.

File: Server/Core/Task.cs [hunk 2]
```diff
@@ -13,73 +14,76 @@ public class Task
...
 
         #endregion
-
+
         #region Properties
-
-        public string Name
-        {
+
+        [Required(ErrorMessage = "Invalid task name.")]
```
Comment (LingForCC) [inline Server/Core/Task.cs]: Why do we need this here?

File: Server/Core/TaskRepository.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Collections.Generic;
+
+namespace Core
+{
+    public class TaskRepository : ITaskRepository
+    {
+        public static List<Task> _store = new List<Task>();
```
Comment (LingForCC) [inline Server/Core/TaskRepository.cs]: Should this be a private field?

File: Server/Core/TaskRepository.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+        public void Add(Task task)
+        {
+            _store.Add(task);
```
Comment (LingForCC) [inline Server/Core/TaskRepository.cs]: Do we need to check null reference here?

File: Server/Core/Task.cs [hunk 5]
```diff
@@ -13,73 +14,76 @@ public class Task
...
-        public bool IsCompleted
-        {
+        public bool IsCompleted
+        {
             get;
-            private set;
+            private set;
         }
 
+        public int Id { get; set; }
```
Comment (LingForCC) [inline Server/Core/Task.cs]: Type of int may not be good enough here. Maybe we should think about using uuid or string.

File: Server/Core/Task.cs (hunk 2 above)
Comment (RukaiYu) [reply Server/Core/Task.cs]: This is a validation checking, actually, it should not goes with the domain model, I've moved it to request model separately.

File: Server/Core/TaskRepository.cs (hunk 3 above)
Comment (RukaiYu) [reply Server/Core/TaskRepository.cs]: yeah, just for testing... :D

File: Server/Core/Task.cs (hunk 5 above)
Comment (RukaiYu) [reply Server/Core/Task.cs]: right

**Comment by RukaiYu** [comment]
@LingForCC  How to make the CI restore the newly added nuget packages? Seems it is failed with this.

File: Server/API/Controllers/TaskController.cs [hunk 6]
```diff
@@ -1,33 +1,64 @@
...
         {
-            //do nothing for now
+            try
+            {
+                var task = _mapper.Map<Task>(request);
+                task.Id = NUlid.Ulid.NewUlid().ToString();
+
+                _taskRepository.Add(task);
+
+                // should we response the task details?
```
Comment (LingForCC) [inline Server/API/Controllers/TaskController.cs]: I think it should be fine at the moment.

File: Server/API/Models/PostNewTaskRequestModel.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace API.Models
+{
+    public class PostNewTaskRequestModel
```
Comment (LingForCC) [inline Server/API/Models/PostNewTaskRequestModel.cs]: Do we need the request model here? It seems that we have the same validation logic in Task Repository as well.

File: Server/Core/TaskRepository.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace Core
+{
+    public class TaskRepository : ITaskRepository
+    {
+        private static List<Task> _store = new List<Task>();
+
+        public void Add(Task task)
+        {
+            if (null == task)
```
Comment (LingForCC) [inline Server/Core/TaskRepository.cs]: Shall we combine the validation logic into a function named 'ValidateTaskName'?

File: Server/CoreTest/TaskTest.cs [hunk 9]
```diff
@@ -12,5 +11,55 @@ public void TestAddDedpendencyWithNull()
...
+            ITaskRepository repo = new TaskRepository();
+
+            repo.Add(task);
+
+            Assert.Equal(repo.FindById(task.Id).Name, task.Name);
+        }
+
+        [Fact]
+        public void TestAddTaskWithInvalidEmptyName()
+        {
```
Comment (LingForCC) [inline Server/CoreTest/TaskTest.cs]: If we have the validation function named ValidateTaskName above, shall we just test against this function? Once this function is tested, can we test against the AddTask function in Controller to make a integration test to make sure the workflow is working?

**Comment by LingForCC** [comment]
I will merge the PR first for ease of development.

File: Server/Core/TaskRepository.cs (hunk 8 above)
Comment (RukaiYu) [reply Server/Core/TaskRepository.cs]: Yeah, Indeed I think we should not place the validation logic in the repository but in service or something else. I'm planning to have a refactoring on the repository pattern

File: Server/CoreTest/TaskTest.cs (hunk 9 above)
Comment (RukaiYu) [reply Server/CoreTest/TaskTest.cs]: Yes

File: Server/API/Models/PostNewTaskRequestModel.cs (hunk 7 above)
Comment (RukaiYu) [reply Server/API/Models/PostNewTaskRequestModel.cs]: I think we need request model (or call it resource model is more accurate in REST API).
For the validations rules, DataAnnotations way is more convenient, but another one is more unit-text friendly. So let's keep the second one (separate it in Task Service)?

**Pull request merged**